### ISSUE TITLE
8. Recurrent Neural Network

### ISSUE THREAD

**Issue by p-jiho** [p-jiho/natural-language#8]
- 피드 포워드 신경망의 한계 : 입력 길이가 고정
   -> 해결 : 순환 신경망(RNN)

## 1. Recurrent Neural Network, RNN
- 입력과 출력을 시퀀스 단위로 처리

#### 1) 순환 신경망(RNN)
- 은닉층에서 활성화 함수를 통해 나온 결과값을 출력층의 방향으로 보내면서 동시에 다시 은닉층 노드의 입력으로 보낸다.
- 메모리 셀 or RNN 셀 : 은닉층에서 활성화 함수를 통해 결과를 내보내려고 한다. 이전의 값을 기억하는 노드 -> 재귀적 활동을 함
- 은닉 상태 : 메모리 셀이 출력층 또는 다음 시점의 자신에게 보내는 값
- 피드 포워드 신경망과 달리 뉴런의 단위보다는 입력벡터, 출력벡터, 은닉상태라는 표현을 사용
- RNN 셀의 각 시점의 입출력 단위 : 보편적으로는 "단위벡터"
- 일 대 다 구조 : 이미지를 통해 제목을 출력하는 이미지 캡셔닝
- 다 대 일 구조 : 감성분류, 스팸 메일 분류 등
- 다 대 다 구조 : 번역기, 개채명 인식, 품사 태깅
- t 시점의 은닉 상태값은 입력층을 위한 가중치와 t-1 시점의 은닉 상태값을 위한 가중치로 총 2개의 가중치가 필요하다.
- 은닉층 연산에는 주로 하이퍼볼릭탄젠트 함수가 사용
- 은닉층이 하나일 경우 모든 시점에서 가중치의 값이 동일하지만 두 개 이상일 경우 각 은닉층의 가중치 값은 다르다.
- 출력층의 활성화 함수 
   -> 이진 분류 : 로지스틱 회귀, 시그모이드 함수
   -> 다중 클래스 분류 : 소프트맥스 회귀, 소프트 맥스 함수

#### 2) Keras로 RNN 구현
- 임포트 : `from tensorflow.keras.layers import SimpleRNN`
- RNN 층 추가 : `model.add(SimpleRNN(hidden_units))`
      - `hidden_units`인자 : 은닉 상태의 크기 정의 output_dim과 동일
      - `input_shape`의 `timestep` 인자 : 입력 시퀀스의 길이
      - `input_shape`의 `input_dim` : 입력의 크기
- RNN 층은 3D 텐서를 입력으로 받는다.
- 출력
    1. 2D 텐서 : 메모리 셀의 최종 시점의 은닉상태만 리턴, batch_size, output_dim
    2. 3D 텐서 : 각 시점의 은닉 산태값을 모아서 전체 시퀀스를 리턴. batch_size, timesteps, output_dim
        -> `return_sequences` 변수를 `True`로 설정하여 사용가능
- `model.add` 함수에서 위 코드처럼 층을 추가하면 batch_size는 알 수없으므로 `None` 이된다.
- batch_size를 추가하려면 `batch_input_shape` 인자를 이용해야한다.

#### 3) 파이썬으로 RNN 구현
- `timesteps` : 시점의 수. 문장의 길이
- `input_dim` : 입력의 차원, 단어 벡터의 차원
- `hidden_units` : 은닉 상태의 크기, 메모리 셀의 용량
- 입력에 대한 가중치의 크기 : (은닉 상태의 크기)와 (입력의 차원)의 곱
- 은닉 상태에 대한 가중치의 크기 : (은닉 상태의 크기)와 (은닉 상태의 크기)의 곱
- 편향의 크기 : (은닉 상태의 크기)

#### 4) Deep Recurrent Neural Network
- 2개 이상의 은닉층을 가지는 경우
- 다음 은닉층이 존재하면 `return_sequences=True`로 지정해주어야한다.

#### 5) Bidirectional Recurrent Neural Network
- 이전 시점의 입력 뿐만 아니라 이후 시점의 입력도 예측에 기여한다.
- 하나의 출력값을 위해 두 개의 메모리 셀을 사용한다.
- 임포트 : `from tensorflow.keras.models import Bidirectional`
- 은닉층 추가 : `model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True), input_shape))`
- 은닉층을 더 추가할 수 있지만 꼭 성능이 좋아지는 것은 아님 -> 학습할 수 있는 양이 증가하지만 훈련 데이터도 많은 양이 필요

<br></br>

## 2. Long Short-Term Memory, LSTM
- 가장 단순한 형태의 RNN을 변형한 것

#### 1) 바닐라 RNN의 한계
- 장기 의존성 문제 : 시점이 길어질 수록 앞의 정보가 충분히 전달되지 못한다.
- 짧은 시퀀스에서만 이전의 계산 결과에 의존하는 모습을 보인다.

#### 2) LSTM
- 메모리 셀에 입력 게이트, 망각 게이트, 출력 게이트 추가하여 필요한 정보와 불필요한 정보를 구분한다.
   -> 3개의 게이트에는 모두 시그모이드 함수 존재 : 0과 1 사이의 값으로 게이트를 조절
- 셀 상태 : 이전 시점의 셀 상태가 다음 시점의 셀 상태를 구하기 위한 입력으로 사용

1. 입력 게이트
- 현재 정보를 기억하기 위한 게이트
- (현재 시점의 입력값과 가중치의 곱) + (이전 시점의 은닉상태와 가중치의 곱) -> 시그모이드 함수
- (현재 시점의 입력값과 가중치의 곱) + (이전 시점의 은닉상태와 가중치의 곱) -> 하이퍼볼릭탄젠트 함수
- 두 개의 값을 가지고 기억할 정보의 양을 구함

2. 삭제 게이트
- 현재 정보를 삭제하기 위한 게이트
- (현재 시점의 입력값과 가중치의 곱) + (이전 시점의 은닉 상태와 가중치의 곱) -> 시그모이드 함수를 지남
- 0과 1 사이의 값이 삭제 과정을 거친 정보의 양 -> 0에 가까울 수록 정보가 많이 삭제

3. 셀 상태
- 입력 게이트에서 구한 두 개의 값에 대해 원소별 곱을 진행
- (입력 게이트에서 선택된 기억) + (삭제 게이트의 결과값) => 다음 시점의 셀로 넘어감
- 삭제 게이트가 0이라면? : 이전 시점의 값이 영향력이 사라져 오직 입력 게이트의 결과만 현재에 영향을 준다.
- 입력 게이트가 0이라면? : 현재 시점의 값이 영향력이 사라져 오직 이전 시점의 셀 상태의 값에만 의존한다.
- 삭제 게이트 : 이전 시점의 입력을 얼마나 반영할 것인가?
- 입력 게이트 : 현재 시점의 입력을 얼마나 반영할 것인가?

4. 출력 게이트와 은닉 상태
- 현재 시점의 은닉 상태를 결정한다.
- 현재 시점의 입력값, 이전 시점의 은닉 상태 -> 시그모이드 함수를 지남

<br></br>

## 3. Gated Recurrent Unit, GRU
- LSTM의 성능은 유지하지만 구조를 간단화시킴

#### 1) GRU(Gated Recurrent Unit)
- 3개의 게이트가 아닌 업데이트 게이트와 리셋 게이트만 존재한다.

#### 2) 케라스에서의 GRU
- `model.add(GRU(hidden_size, input_shape=(timesteps, input_dim)))`

<br></br>

#### 1) 임의의 입력 생성
-  RNN은 3D 텐서를 입력받으므로 `shape` 함수를 사용했을 때 3차원의 크기를 가지도록 입력을 생성해야한다.

#### 2) SimpleRNN 이해하기
```
from tensorflow.keras.layers import SimpleRNN
rnn = SimpleRNN(3) # 3은 은닉 상태의 크기를 말한다.
rnn(train_X)
```
- `SimpleRNN` 함수의 `return_sequences` 인자를 `True`로 지정한다면?
   : 모든 시점에 대한 은닉 상태의 값을 출력하게 된다.
- `SimpleRNN` 함수의 `return_state` 인자를 `True`로 지정한다면?
   : 두 개의 출력을 리턴하여 두 개의 변수로 값을 받아야한다. 첫 번째는 모든 시점의 은닉 상태, 두 번째는 마지막 시점의 은닉 상태이다.
- `return_sequences` 인자는 `False`이고, `return_state` 인자는 `True`로 지정된다면?
  : 두 개의 출력 모두 마지막 시점의 은닉 상태를 출력하게 된다.

#### 3) LSTM 이해하기
```
from tensorflow.keras.layers import LSTM
lstm = LSTM(3)
lstm(train_X)
```
- `return_sequences`를 `False`, `return_state`가 True`로 지정한다면?
  : 3개의 결과를 반환한다. 마지막 시점의 은닉상태, 마지막 시점의 은닉상태, 마지막 시점의 셀 상태
- `return_sequences`를 `True`, `return_state`가 True`로 지정한다면?
  : 모든 시점의 은닉상태, 마지막 시점의 은닉상태, 마지막 시점의 셀 상태

#### 4) Bidirectional(LSTM) 이해하기
```
from tensorflow.keras.layers import Bidirectinal
# return_sequences 값에 따른 결과 비교를 위해 은닉 상태의 값을 고정
k = tf.keras.initializers.Constant(value=0.1)
b = tf.keras.initializers.Constant(value=0)
r = tf.keras.initializers.Constant(value=0.1)
bilstm = Bidirectional(LSTM(3, kernel_initalizer=k, bias_initializer=b, recurrent_initializer=r)
bilstm(train_X)
```
- `return_sequences`는 `False`, `return_state`는 `True`로 지정한다면?
  : 5개의 값을 반환한다.  마지막 시점의 은닉상태, 정방향 LSTM의 은닉 상태와 셀상태, 역방향 LSTM의 은닉상태와 셀 상태를 반환한다. 정방향 LSTM의 마지막 시점의 은닉 상태와 역방향 LSTM의 첫번째 시점의 은닉 상태가 연결되어 다음 층으로 연결된다.
- `return_state`는 `True`, `return_state`는 `True`로 지정한다면?
  : `False`였을 때와 달리 모든 은닉 상태가 출력이 된다. 또 정방향 LSTM의 마지막 시점의 은닉 상태는 역방향 LSTM의 마지막 시점의 은닉 상태와 연결이 된다.

<br></br>

#### 1) RNNLM
- n-gram 언어 모델과 NNLM : 고정된 개수의 단어를 입력해야한다.
- RNN : 시점 개념이 도입되면서 입력 길이를 고정하지 않을 수 있다.
- RNNLM : RNN으로 만든 언어 모델
- 이전 시점의 출력을 현재 시점의 입력으로 한다.
- 테스트 과정에서 각 시점은 이전 시점들에서 나온 시퀀스로 인해 결정이 된다.
- 훈련 과정에서는 시퀀스를 넣으면 각 시점의 레이블들이 정해져있어 예측하도록 훈련하는 것이다.
- 교사 강요 : 테스트 과정에서 이전 시점의 출력이 현재 시점의 입력으로 사용하는 RNN 모델. 이전 시점의 예측값을 입력값으로 사용하는 것이 아닌 이전 시점의 레이블을 입력값으로 사용한다.
- 이전 시점의 예측값을 사용하면 한 번 예측이 잘못되면 뒤까지 영향을 미친다. -> 교사 강요를 사용해 빠르고 효율적으로 훈련시킨다.
- 출력층에서 소프트맥스 함수, 오차 계산을 위해 손실함수로 크로스 엔트로피 함수를 사용한다.

순서
1. 입력 단어의 원 핫 벡터를 입력 받는다.
2. 임베딩층을 지난다. : NNLM의 투사층과 동일하다. 룩업 테이블을 수행해 임베딩 벡터를 얻는다. 임베딩 행렬의 크기는 (단어집합의 크기)와 (임베딩 벡터의 크기)의 곱이다.
3. 은닉층 : 받은 임베딩 벡터는 이전 시점의 은닉 상태와 연산을 하여 은닉 상태를 계산한다.
4. 출력층 : 소프트 맥스 함수를 지나 예측값을 구한다.
5. 원-핫 벡터의 값과 가까워져야한다. 손실함수로 크로스 엔트로피 함수를 사용한다.
6. 역전파 : 가중치 행렬들이 학습되며 임베딩 벡터값들도 학습이 된다.

<br></br>

## 6. RNN을 이용한 택스트 생성
- 다 대 일 구조의 RNN 예시이다.

#### 1) RNN을 이용해 텍스트 생성
- 여러 문장을 가지고 데이터를 재구성한다.
- "밥을 먹고 있다" 라는 데이터를 재구성하면
    - sample 1 : X - 밥을 , y - 먹고
    - sample 2 : X - 밥을 먹고, y - 있다

1. 데이터에 대한 이해와 전처리
- 단어 집합 생성
```
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer()
tokenizer.fit_on_texts([text])
```
: 단어 집합의 크기는 모든 길이를 동일하게 만들어주기 위해 사용할 인덱스를 고려하여 즉, 패딩 작업을 생각하여 +1을 해준다.

- 훈련 데이터를 만든다. : `text.split` 함수로 문장들을 나눠주고 `tokenizer.texts_to_sequences` 함수를 이용해 문장을 토큰화 한다. 그리고 `for` 문을 통해 토큰화 한 단어를 [1],[1,2][1,2,3]과 같이 문장 마다 단어를 하나씩 늘려가는 훈련 데이터를 만든다.
- 훈련 데이터는 레이블로 사용될 단어들이 분리되지 않았다. 즉, [밥을, 먹고],[밥을, 먹고, 있다]로 구성이 되어있다.

- 가장 긴 샘플 길이를 기준으로 전체 샘플의 길이를 일치시킨다. 
```
from tensorflow.keras.preprocessing.sequence import pad_sequences
sequences = pad_sequences(sequences, maxlen = max_len, padding="pre")
```
: 패딩 작업을 한다. 앞을 0으로 채운다.

- 레이블을 분리한다. : `np.array(sequences)`를 이용해 데이터를 만들고 마지막 줄만 분리를 한다.

- 레이블에 대한 원-핫 인코딩 수행
```
from tensorflow.keras.utils import to_categorical
y = to_categorical(y, num_classes=vocab_size)
```
: `y`는 레이블 데이터, `vocab_size`는 단어집합의 크기로 지정하여 수행한다.

#### 2) 모델 설계
- 임베딩 벡터 차원은 10으로 설정한다.
```
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
embedding_dim = 10
model = Sequential()
model.add(Embedding(vocab_size, embedding_dim))
```

- 은닉 상태 크기는 32로 지정하고 다 대 일 구조의 RNN을 사용한다.
```
from tensorflow.keras.layers import SimpleRNN
hidden_units = 32
model.add(SimpleRNN(hidden_units))
```

- 마지막 시점에서 다중 클래스 분류 문제를 수행하는 경우 출력층에 소프트 맥스 회귀를 사용한다.
```
from tensorflow.keras.layers import Dense
model.add(Dense(vocab_size, activation="softmax"))
```
 
- 손실 함수로 크로스 엔트로피 함수를 사용한다.
`model.compile(loss = "categorical_crossentropy",optimizer = "adam", metrics =["accuracy"])`

- 200 epoch를 수행한다.
`model.fit(X, y, epochs = 200, verbose=2)`

- 예측 함수 생성
   - `tokenizer.texts_to_sequences`를 이용해 정수 인코딩
   - 정수 인코딩한 결과를 `pad_sequences`를 이용해 패딩 작업
   - 패딩 작업한 결과를 `model.predict`를 이용해 결과 예측
   - 예측한 결과를 `np.argmax`를 이용해 해당 단어의 인덱스 추출 : 해당 함수는 가장 큰 값을 가지는 수의 인덱스를 추출한다.
   - `for`문과 `tokenizer.word_index.items()`를 이용해 동일한 인덱스를 가진 단어를 추출
   - `for`문을 이용해 예측한 단어들을 추가하며 다음 단어를 하나씩 예측

#### 2) LSTM을 이용해 텍스트 생성
- 뉴스 기사 제목 데이터이다. 기사 제목을 예측하는 코드이다.
- 데이터 학습 과정은 위와 동일하고 데이터 전처리 과정이 포함되어있으므로 https://wikidocs.net/45101 이 링크에서 자세한 코드를 확인하는 것이 좋다.

<br></br>

## 7. 문자 단위 RNN
- 현재까지는 입력과 출력 단위가 단어 벡터이다.
- 단어 벡터가 아닌 문자 레벨로 변경하여 구현이 가능하다.

#### 1) 문자 단위 RNN 언어 모델
- 임베딩층을 사용하지 않는다.

1. 데이터에 대한 이해와 전처리
- 고전 소설 데이터
  : `import urllib.request`을 임포트해 `urllib.request.urlretrieve` 함수를 이용해 데이터를 로드하고 `open` 함수를 이용해 데이터를 읽어온다.
- `strip()` 함수를 통해 \r과 \n 제거.
- `lower()` 함수를 통해 소문자화
- `decode("ascii","ignore")` 함수를 통해 바이트 열 제외
- `[]`를 이용해 생성된 빈 리스트에 `append` 함수를 이용해 읽어들인 문장을 원소로 넣는다.
- 문장이 의미있게 나누어지지 않았으므로 `' '.join`함수를 이용해 문장들을 통합한다.
- `sorted(list(set(data)))` 함수를 이용해 `data`를 문자 집합으로 만든다.
- 단어를 표현하기 위한 문자 수는 적으므로 문자 집합의 크기는 작다.
- 정수 인코딩 : `enumerate` 함수를 이용해 정수를 부여한다.
- 샘플의 길이을 정해 전체 문자열로 나누어 샘플 수를 구할 수 있다.
- 정수 인코딩, 원-핫 인코딩을 한다.

2. 모델 설계
- 다 대 다 구조의 LSTM을 사용한다.
```
model.add(LSTM(hidden_units, input_shape = (None, train_X.shape[2]), return_sequences=True))
model.add(TimeDistributed(Dense(vocab_size, activation="softmax")))
```
- 임베딩층을 사용하지 않는다. 
- `TimeDistributed` 함수는 각 스텝마다 오류를 계산해 반영하는 것이다.

---

- 참고
>> 딥 러닝을 이용한 자연어 처리 입문, 유원준, 안상준
      * 위 글 내용은 이 책의 내용을 공부하여 정리한 것입니다.
